File: hospital-ai/frontend/src/types/diagnosis.ts
/** Diagnosis Agent types — clinical decision support (assists, never replaces, a physician). */

export type SeverityLevel = 'Very Low' | 'Low' | 'Moderate' | 'High' | 'Critical';

export const SEVERITY_LEVEL_TONE: Record<
  string,
  'green' | 'blue' | 'amber' | 'red' | 'gray'
> = {
  'Very Low': 'green',
  Low: 'green',
  Moderate: 'amber',
  High: 'red',
  Critical: 'red',
};

export interface SymptomCluster {
  body_system: string;
  symptoms: string[];
  duration_hint?: string | null;
  severity_hint?: string | null;
  related_conditions: string[];
}

export interface SymptomAnalysis {
  clusters: SymptomCluster[];
  total_symptoms: number;
  notable_vitals: Array<Record<string, unknown>>;
  notable_labs: Array<Record<string, unknown>>;
  narrative: string;
}

export interface DifferentialDiagnosis {
  condition: string;
  confidence: number;
  supporting_symptoms: string[];
  supporting_labs: string[];
  supporting_history: string[];
  contradicting_evidence: string[];
  recommended_specialists: string[];
  body_system?: string | null;
}

export interface DiseaseProbability {
  condition: string;
  probability_pct: number;
  confidence: number;
  evidence_used: string[];
  risk_contribution: number;
  risk_category?: string | null;
}

export interface SeverityAssessment {
  level: string;
  score: number;
  explanation: string;
  contributing_factors: string[];
}

export interface TreatmentPath {
  recommended_specialists: string[];
  recommended_department?: string | null;
  diagnostic_tests: string[];
  imaging: string[];
  urgency: string;
  notes: string;
}

export interface ClinicalDecisionSupport {
  possible_diagnoses: string[];
  supporting_evidence: string[];
  suggested_tests: string[];
  risk_factors: string[];
  relevant_history: string[];
  clinical_notes: string[];
  disclaimer: string;
}

export interface DiagnosisResult {
  id: string;
  patient_id: string;
  processing_job_id?: string | null;
  chief_complaint?: string | null;
  focus_symptoms_json: string[];
  symptom_analysis_json: SymptomAnalysis;
  differential_diagnoses_json: DifferentialDiagnosis[];
  probability_scores_json: DiseaseProbability[];
  severity_assessment_json: SeverityAssessment;
  treatment_path_json: TreatmentPath;
  clinical_decision_support_json: ClinicalDecisionSupport;
  summary?: string | null;
  engine: string;
  status: string;
  error_message?: string | null;
  processing_time_ms?: number | null;
  created_at: string;
  updated_at?: string | null;
}

export interface DiagnosisStartResult {
  patient_id: string;
  status: string;
  processing_time_ms: number;
  summary: string;
  engine: string;
  warnings: string[];
  symptom_analysis: SymptomAnalysis;
  differential_diagnoses: DifferentialDiagnosis[];
  probability_scores: DiseaseProbability[];
  severity_assessment: SeverityAssessment;
  treatment_path: TreatmentPath;
  clinical_decision_support: ClinicalDecisionSupport;
  diagnosis_result: DiagnosisResult;
}

export interface DiagnosisHistoryItem {
  id: string;
  created_at: string;
  summary?: string | null;
  severity_level?: string | null;
  top_condition?: string | null;
  status: string;
}

File: hospital-ai/backend/app/schemas/diagnosis.py
"""Pydantic schemas for the Diagnosis Agent."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, Field


class DiagnosisStartRequest(BaseModel):
    patient_id: UUID
    chief_complaint: Optional[str] = None
    focus_symptoms: List[str] = Field(default_factory=list)


class SymptomClusterOut(BaseModel):
    body_system: str
    symptoms: List[str] = Field(default_factory=list)
    duration_hint: Optional[str] = None
    severity_hint: Optional[str] = None
    related_conditions: List[str] = Field(default_factory=list)


class SymptomAnalysisOut(BaseModel):
    clusters: List[SymptomClusterOut] = Field(default_factory=list)
    total_symptoms: int = 0
    notable_vitals: List[Dict[str, Any]] = Field(default_factory=list)
    notable_labs: List[Dict[str, Any]] = Field(default_factory=list)
    narrative: str = ""


class DifferentialDiagnosisOut(BaseModel):
    condition: str
    confidence: float = 0.0
    supporting_symptoms: List[str] = Field(default_factory=list)
    supporting_labs: List[str] = Field(default_factory=list)
    supporting_history: List[str] = Field(default_factory=list)
    contradicting_evidence: List[str] = Field(default_factory=list)
    recommended_specialists: List[str] = Field(default_factory=list)
    body_system: Optional[str] = None


class DiseaseProbabilityOut(BaseModel):
    condition: str
    probability_pct: float = 0.0
    confidence: float = 0.0
    evidence_used: List[str] = Field(default_factory=list)
    risk_contribution: float = 0.0
    risk_category: Optional[str] = None


class SeverityAssessmentOut(BaseModel):
    level: str = "Very Low"
    score: float = 0.0
    explanation: str = ""
    contributing_factors: List[str] = Field(default_factory=list)


class TreatmentPathOut(BaseModel):
    recommended_specialists: List[str] = Field(default_factory=list)
    recommended_department: Optional[str] = None
    diagnostic_tests: List[str] = Field(default_factory=list)
    imaging: List[str] = Field(default_factory=list)
    urgency: str = "Routine"
    notes: str = ""


class ClinicalDecisionSupportOut(BaseModel):
    possible_diagnoses: List[str] = Field(default_factory=list)
    supporting_evidence: List[str] = Field(default_factory=list)
    suggested_tests: List[str] = Field(default_factory=list)
    risk_factors: List[str] = Field(default_factory=list)
    relevant_history: List[str] = Field(default_factory=list)
    clinical_notes: List[str] = Field(default_factory=list)
    disclaimer: str = ""


class DiagnosisResultOut(BaseModel):
    id: UUID
    patient_id: UUID
    processing_job_id: Optional[UUID] = None
    chief_complaint: Optional[str] = None
    focus_symptoms_json: List[str] = Field(default_factory=list)
    symptom_analysis_json: Dict[str, Any] = Field(default_factory=dict)
    differential_diagnoses_json: List[Any] = Field(default_factory=list)
    probability_scores_json: List[Any] = Field(default_factory=list)
    severity_assessment_json: Dict[str, Any] = Field(default_factory=dict)
    treatment_path_json: Dict[str, Any] = Field(default_factory=dict)
    clinical_decision_support_json: Dict[str, Any] = Field(default_factory=dict)
    summary: Optional[str] = None
    engine: str = "rule_based"
    status: str = "Completed"
    error_message: Optional[str] = None
    processing_time_ms: Optional[int] = None
    created_at: datetime
    updated_at: Optional[datetime] = None


class DiagnosisStartResponse(BaseModel):
    patient_id: UUID
    status: str = "Completed"
    processing_time_ms: int
    summary: str
    engine: str
    warnings: List[str] = Field(default_factory=list)
    symptom_analysis: SymptomAnalysisOut
    differential_diagnoses: List[DifferentialDiagnosisOut] = Field(default_factory=list)
    probability_scores: List[DiseaseProbabilityOut] = Field(default_factory=list)
    severity_assessment: SeverityAssessmentOut
    treatment_path: TreatmentPathOut
    clinical_decision_support: ClinicalDecisionSupportOut
    diagnosis_result: DiagnosisResultOut


class DiagnosisHistoryItemOut(BaseModel):
    id: UUID
    created_at: datetime
    summary: Optional[str] = None
    severity_level: Optional[str] = None
    top_condition: Optional[str] = None
    status: str = "Completed"
